spice/spice.py

```diff
@@ -371,25 +382,51 @@ async def get_response(
             If the provider/model does not support response_format, this argument will be ignored.
 
             name: If given, will be given this name when logged.
+            validator: If given, will be called with the text of the response. If it returns False, the response will be discarded and another attempt will be made.
+            streaming_callback: If given, will be called with the text of the response as it is received.
+            retries: The number of times to retry getting a valid response. If 0, will not retry.
         """
-
-        text_model = self._get_text_model(model)
-        client = self._get_client(text_model, provider)
-        call_args = self._fix_call_args(messages, text_model, False, temperature, max_tokens, response_format)
-
         start_time = timer()
-        with client.catch_and_convert_errors():
-            chat_completion = await client.get_chat_completion_or_stream(call_args)
-        end_time = timer()
-        text, input_tokens, output_tokens = client.extract_text_and_tokens(chat_completion)
-
-        cost = text_request_cost(text_model, input_tokens, output_tokens)
-        if cost is not None:
-            self._total_cost += cost
+        cost = 0
+        for i in range(retries + 1):
+            text_model = self._get_text_model(model)
+            client = self._get_client(text_model, provider)
+            call_args = self._fix_call_args(
+                messages, text_model, streaming_callback is not None, temperature, max_tokens, response_format
```

Comment:
I think if people want to use a streaming_callback they should just call stream_response; I don't think it makes sense for this function to ever make a streaming call

Reply:
We're constantly getting a streaming response just to print it ourselves in butler. It seems like if you only care about streaming for printing somewhere handling it in spice makes client code a lot cleaner. Also I'm not sure how we'd handle `streaming_response` with validation since when it returns it doesn't yet know if validation will succeed. Could handle it in StreamingSpiceResponse but that gets complicated.

Reply:
Ah that does make sense. This looks good then